framework/include/constraints/MortarConstraintBase.h

```diff
@@ -145,9 +170,18 @@ class MortarConstraintBase : public Constraint,
   /// the normals along the primary face
   const MooseArray<Point> & _normals_primary;
 
-  /// the tangents along the secondary face
+  /// Tangent vectors on the secondary faces (libmesh)
   const MooseArray<std::vector<Point>> & _tangents;
 
+  /// Tangents vector on the secondary faces (householder from normal vectors)
+  std::array<std::vector<Point>, 2> _tangents_3d;
+
+  /// Map to locate nodes from higher dimensional element to lower dimensional elements
+  std::map<unsigned int, unsigned int> _secondary_ip_lowerd_map;
+
+  /// Map to locate nodes from higher dimensional element to lower dimensional elements
+  std::map<unsigned int, unsigned int> _primary_ip_lowerd_map;
```

Comment:
Ordinarily I would suggest `std::unordered_map` but these containers should be really small and a binary search in a `std::map` may be faster